### ISSUE TITLE
Use Shape.zigZagTo

### ISSUE THREAD

**Issue by samreid** [phetsims/faradays-law#153]
Faraday's law has this zig-zag code:

```js
    // Create the filament, which is a zig-zag shape.
    var filamentShape = new Shape().moveToPoint( filamentTopPoint );
    for ( var i = 0; i < NUM_FILAMENT_ZIG_ZAGS - 1; i++ ) {
      var yPos = filamentTopPoint.y + ( filamentBottomPoint.y - filamentTopPoint.y ) / NUM_FILAMENT_ZIG_ZAGS * ( i + 1 );
      var xZig = i % 2 === 0 ? FILAMENT_ZIG_ZAG_SPAN : 0;
      filamentShape.lineTo( filamentTopPoint.x + xZig, yPos );
    }
```

Now that  `Shape.zigZagTo` was added in https://github.com/phetsims/kite/issues/75, it may be better for this simulation to use it instead.

**Comment by jbphet** [comment]
I gave this a shot, but it looks like there may be an issue with `Shape.zigZagToPoint` not handling non-integer values for `numberZigZags`.  Here is what the shape looked like when using a value of 4:

![image](https://user-images.githubusercontent.com/3082831/56935005-c3e49480-6aab-11e9-88bf-20c99ef7cd14.png)

This isn't to bad except that it looks less symmetric than the current implementation, so I tried adding another half cycle with  `numberZigZags` = 4.5 and go this:

![image](https://user-images.githubusercontent.com/3082831/56935226-f2af3a80-6aac-11e9-9be1-88b8adb3c030.png)

This doesn't look right.  It seem like a half cycle should end at the midpoint.

Assigning back to @samreid to determine if there is a bug in this and to follow up if so.  If only integer values are intended to be handled, an assert should be added to make this clear.

**Comment by samreid** [comment]
The current implementation requires an integer.  @Denz1994 ran into the same problem or a similar problem in https://github.com/phetsims/capacitor-lab-basics/issues/258#issuecomment-486489715 and is looking into it.

**Comment by samreid** [comment]
The proposed API is:
```js
zigZagToPoint: function( endPoint, amplitude, numberZigZags, symmetrical )
```

But I'm wondering if it would serve us better to use a more wave-oriented API, like 

```js
zigZagToPoint: function( endPoint, amplitude, wavelength, phase )
```

Then we can have a single implementation for symmetrical/asymmetrical zig-zags.  But on the other hand, if you want to specify 7 zig zags from the start point, you need to use the Shape.currentPoint.distanceTo(endPoint) to compute the wavelength.

Additionally, we may at some point want a "sineWaveToPoint" function, and it would be good to have the same API.

@Denz1994 what do you think?

**Comment by samreid** [comment]
@Denz1994 and I discussed this in slack, and decided we would like to review the proposed API with the development team (see preceding comment).  After the API is decided, then we can schedule implementation.  @Denz1994 and I noted that this is an isolated "bite size" issue that may lend itself to development by an intern or student developer.

**Comment by samreid** [comment]
I realized the latter proposed API in https://github.com/phetsims/faradays-law/issues/153#issuecomment-489801927 is overconstrained.  The following variables are not mutually exclusive: start point, end point, phase, wavelength.  If you specify start point, phase and wavelength, that will determine the amplitude near the end point, so we will have to find an alternative.  This would be a good problem for @veillette or @jonathanolson to look into.

**Comment by veillette** [comment]
The current implementation with the symmetrical flag makes an assumption that the zig-zags are vertically aligned which is not warranted.

**Comment by veillette** [comment]
Mathematically, this would be preferable
````
    /**
     * Zig-zags between the current point and the specified point.
     * Implementation moved from circuit-construction-kit-common on April 22, 2019.
     * @public
     *
     * @param {Vector2} endPoint - the end of the shape
     * @param {number} amplitude - the vertical amplitude of the zig zag wave, signed to choose initial direction
     * @param {number} numberZigZags - the number of complete oscillations
     * @param {boolean} symmetrical - flag for drawing a symmetrical zig zag
     * @public
     */
    zigZagToPoint: function( endPoint, amplitude, numberZigZags, symmetrical ) {
      this.ensure( endPoint );
      const startPoint = this.getLastPoint();
      const delta = endPoint.minus( startPoint );
      const directionUnitVector = delta.normalized();
      const amplitudeNormalVector = directionUnitVector.perpendicular.times( amplitude );
      const wavelength = delta.magnitude / numberZigZags;

      // Handles cases where zig zag symmetry is required.
      if ( symmetrical ) {
        const wavelength = delta.magnitude / ( numberZigZags+0.5 );
        for ( let i = 0; i < numberZigZags+1; i++ ) {
          const waveOrigin = directionUnitVector.times( i * wavelength ).plus( startPoint );
          const topPoint = waveOrigin.plus( directionUnitVector.times( wavelength / 4 ) ).plus( amplitudeNormalVector );
          const bottomPoint = waveOrigin.plus( directionUnitVector.times( 3 * wavelength / 4 ) ).minus( amplitudeNormalVector );
          this.lineToPoint( topPoint );
          // exclude the last bottom point for the zag
          if ( i < numberZigZags) {
            this.lineToPoint( bottomPoint );
          }
        }
      }
      else {
        for ( let i = 0; i < numberZigZags; i++ ) {
          const waveOrigin = directionUnitVector.times( i * wavelength ).plus( startPoint );
          const topPoint = waveOrigin.plus( directionUnitVector.times( wavelength / 4 ) ).plus( amplitudeNormalVector );
          const bottomPoint = waveOrigin.plus( directionUnitVector.times( 3 * wavelength / 4 ) ).minus( amplitudeNormalVector );
          this.lineToPoint( topPoint );
          this.lineToPoint( bottomPoint );
        }
      }
      return this.lineToPoint( endPoint );
    },
```

There is a lot of duplication in this code that would need to be cleaned up. 
In addition, the current API, only allow for a positive zig, the only way to get a negative zig is to specify a negative amplitude, which is cumbersome and obscure. However, this implementation would handle zig-zags along a diagonal line.

**Comment by veillette** [comment]
The previous commit handles zig-zags for no-vertical orientation using the same API. Assigning to @samreid to review.

**Comment by veillette** [comment]
Here is a screenshot from Faraday's Law on Master where the number of Zig-zags is 8 and symmetric.

![image](https://user-images.githubusercontent.com/7633344/57736321-80765280-7675-11e9-9a79-d331de03554c.png)

**Comment by veillette** [comment]
Rather I should assign to @jbphet for review.

**Comment by pixelzoom** [comment]
5/16/19 dev meeting:

@chrisklus pointed out that there are CT errors related to this that have been occurring for awhile. Should someone address this?

@veillette generalized to handle any direction with integer number of zigzags, see above. This is probably triggering assertions in CT, he will investigate.

@samreid will look at CCK. 

We looks at the API in Shape `zigZagToPoint`, the only question was whether `numberZigZags` needs to be integer.  @jonathanolson your opinion?

**Comment by veillette** [comment]
The CT errors came from CCK were the number of zigZag were not restricted to integer values.
A patch was committed  to address the CT errors (https://github.com/phetsims/circuit-construction-kit-common/commit/c2b62768e29ab4a9e91007628dede557c8deed9c). 
All the other instances of Shape.zigZagTo were already constrained to an integer value for numberZigZags.

CT is now passing.

**Comment by samreid** [comment]
1.1.17 (published) looks like this:
![image](https://user-images.githubusercontent.com/679486/58896059-77940200-86b2-11e9-9005-02df0f823dba.png)

But master looks like this:
![image](https://user-images.githubusercontent.com/679486/58896081-7e227980-86b2-11e9-9f91-d7e46e6f9af8.png)

It seems there are too many zig-zags now.  I'll mark this for blocking next publication.

**Comment by jonathanolson** [comment]
> We looks at the API in Shape zigZagToPoint, the only question was whether numberZigZags needs to be integer. @jonathanolson your opinion?

I can imagine having fractional numberZigZags personally (if you wanted one end to be a peak and the other a valley, for instance), but it sounds unlikely to be needed, and like we probably wouldn't need the complexity.

**Comment by samreid** [comment]
I reduced the number of waves and the amplitude to more closely match master.  Reassigning to responsible dev @jbphet to review and see if it is acceptable and if there is anything else for this issue.

**Comment by veillette** [comment]
Unassigning myself

**Comment by jbphet** [comment]
The code looks reasonable, though I reformatted it to meet line length conventions.  There is a bit of a visual difference versus the published version, but I slightly prefer the new version.  I did, however, try to use the new `Shape.zigZagToPoint` method to exactly match the published version to avoid an unnecessary visual change the next time this is published from master, but the new method can't match the old look precisely due to the particular way that the original code worked.  I'm okay with this change.

I took a quick look at `Shape.zigZagToPoint` and have a couple of minor comments:
+ It's marked public twice
+ I don't think the history comment is helpful, we could get that from git history if needed

I did **not** do a thorough review of the new method, and I'm not sure if anyone has, so I'm assigning back to @samreid to address minor issues above and decide if any more review of the method in `Shape` is needed.  He can close after that.

**Comment by samreid** [comment]
Thanks @jbphet, I removed the extraneous `@public` annotation.  Regarding the history comment: we now have (a) a git commit message indicating where this code came from and (b) a comment in the history indicating where this code came from.  At previous developer meetings we decided to add a code comment when code is moved from one repo to another, I don't know whether this decision is to keep it at the head of master forever or if it is sufficient for it to be at the initial move.  I do see other occurrences in master like: https://github.com/phetsims/griddle/blob/f8f4436baedf704306651319a5142c0ae79bbcbd/js/ScrollingChartNode.js#L11

and 

https://github.com/phetsims/axon/blob/9229be637f6ade17061685f23d3e525053f7e542/js/timer.js#L12

Also, do we need to regenerate screenshots for this sim?
